packages/sd-jwt/src/verify.ts

```diff
@@ -0,0 +1,113 @@
+import * as jwt from "jsonwebtoken";
+import * as jose from "jose";
+import * as jwks from "jwks-rsa";
+import { SdJwt } from "./types";
+
+type VerifyOptions = { jwksUri: string };
+type VerifyResult = SdJwt;
+
+function isSupportedAlgorithm(alg: string): alg is jwt.Algorithm {
+  return [
+    "HS256",
+    "HS384",
+    "HS512",
+    "RS256",
+    "RS384",
+    "RS512",
+    "ES256",
+    "ES384",
+    "ES512",
+    "PS256",
+    "PS384",
+    "PS512",
+  ].includes(alg);
+}
+
+// Check a jwt has the attributes defined in the SD-JWT specification
+// It only support first level of disclosure for now
+function jwtIsSdJwt(j: jwt.Jwt): j is SdJwt {
+  // every condition expect this, we can short-circuit the check
+  if (typeof j.payload !== "object") {
+    return false;
+  }
+
+  const hasValidSd =
+    "_sd" in j.payload &&
+    Array.isArray(j.payload["_sd"]) && // eslint-disable-line @typescript-eslint/dot-notation
+    j.payload["_sd"].every((s) => typeof s === "string"); // eslint-disable-line @typescript-eslint/dot-notation
+  const hasValidSdAlg =
+    "_sd_alg" in j.payload && typeof j.payload["_sd_alg"] === "string"; // eslint-disable-line @typescript-eslint/dot-notation
+  const hasNotSdAlg = !("_sd_alg" in j.payload);
+
+  return hasValidSd && (hasValidSdAlg || hasNotSdAlg);
+}
+
+/**
+ * Verify a JWT is structurally valid and its signature is verified.
+ * It returns the parsed JWT object.
+ * It throws an error if the validation fails
+ *
+ * @param token the encoded JWT token to be verified
+ * @param options
+ * @param options.jwksUri URI of the public endpoint of the emitter
+ *
+ * @throws a verification error
+ * @returns the parsed token object
+ */
+async function verifyJWT(
+  token: string,
+  { jwksUri }: VerifyOptions
+): Promise<jwt.Jwt> {
+  // parse header to extract signing informations
+  const { kid, alg } = jose.decodeProtectedHeader(token);
+
+  // retrieve the signing key from the public JWK endpoint
+  const signingKey = await new jwks.JwksClient({
+    cache: true,
+    jwksUri,
+    rateLimit: false,
+  }).getSigningKey(kid);
+
+  // check that the hashing algorithm is defined and supported
+  if (!alg || !isSupportedAlgorithm(alg)) {
```

Comment:
why isn't this check done before calling `jwks.JwksClient`?

Reply:
Good point, done in https://github.com/pagopa/io-wallet/pull/4/commits/e85de0e83f29775d2852b7330842c4a91abb227d